### PR TITLE
Make ReactiveClientHeadersFactory replace headers instead of adding

### LINKED ISSUE #24054: ReactiveClientHeadersFactory ignores rest-client @Produces headers
### Describe the bug

When i use `@RegisterClientHeaders(ReactiveHeaderFactory.class) `
`public class ReactiveHeaderFactory extends  ReactiveClientHeadersFactory`

The Annotation `@Produces(MediaType.APPLICATION_OCTET_STREAM) `
in rest-client is ignored and expected headers are not injected.

The detail of error:
1. Endpoint exposed with Resteasy that consumes application/json:
    `@GET
      @Consumes(MediaType.APPLICATION_JSON)
    `
2. After i execute rest-client(with reactive-headers) call where i cannot see 
expected headers injected by `@Produces(MediaType.APPLICATION_OCTET_STREAM) `

If i remove `@RegisterClientHeaders(ReactiveHeaderFactory.class) ` the headers are injected as exptected.

I have created the sample project where you can simulate the issue (to speed-up you can inside the project [mockoon](https://mockoon.com/) environment and [postman](https://www.postman.com/) collection)

### Expected behavior

When i use `ReactiveClientHeadersFactory` with rest-client
i expect that `@Produces(MediaType.APPLICATION_OCTET_STREAM)`
will include necessary headers

### Actual behavior

When i use `ReactiveClientHeadersFactory` with rest-client
i expect that `@Produces(MediaType.APPLICATION_OCTET_STREAM)`
will include necessary headers but they are not present

### How to Reproduce?

Checkout project https://github.com/herasimau/quarkus-reactiv-header-problem

Import postman collection
Import mockoon environment 
Start quarkus

Run postman end-points:

- http://localhost:9084/header/without-reactive-headers - WORKS
- http://localhost:9084/header/with-reactive-headers - DOESN'NT WORKS

### Output of `uname -a` or `ver`

Darwin Leanids-MacBook-Pro.local 21.3.0 Darwin Kernel Version 21.3.0: Wed Jan  5 21:37:58 PST 2022; root:xnu-8019.80.24~20/RELEASE_ARM64_T6000 arm64

### Output of `java -version`

openjdk version "11.0.14" 2022-01-18 OpenJDK Runtime Environment GraalVM CE 22.0.0.2 (build 11.0.14+9-jvmci-22.0-b05)

### GraalVM version (if different from Java)

_No response_

### Quarkus version or git rev

2.7.2.Final

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Apache Maven 3.8.1 (05c21c65bdfed0f71a2f2ada8b84da59348c4c5d) Maven home: /opt/homebrew/Cellar/maven/3.8.1

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by michalszynkiewicz**
fixes #24054

### CODE REVIEW

**Comment by michalszynkiewicz** [comment]
CC @Sgitario

File: extensions/resteasy-reactive/rest-client-reactive/runtime/src/main/java/io/quarkus/rest/client/reactive/runtime/MicroProfileRestClientRequestFilter.java [hunk 1]
```diff
@@ -72,6 +72,7 @@ public void filter(ClientRequestContext requestContext) {
                 reactiveRequestContext.suspend();
                 MultivaluedMap<String, String> outgoingHeaders = incomingHeaders;
                 reactiveClientHeadersFactory.getHeaders(incomingHeaders).subscribe().with(newHeaders -> {
+                    outgoingHeaders.clear();
```
Comment (geoand) [inline extensions/resteasy-reactive/rest-client-reactive/runtime/src/main/java/io/quarkus/rest/client/reactive/runtime/MicroProfileRestClientRequestFilter.java:75]: Can you explain why this is needed? Prefereably in a source-code comment so it's persistent :)

**Review by Sgitario — COMMENTED** [review]
Is this how is working the resteasy classic vendor of the reactive client header annotation?

If not, I'm not sure if this should be the right behavior. Mostly because of the scenario of using this annotation in sub-resources where the headers of the latest resource will take precedent.

File: extensions/resteasy-reactive/rest-client-reactive/runtime/src/main/java/io/quarkus/rest/client/reactive/runtime/MicroProfileRestClientRequestFilter.java (hunk 1 above)
Comment (michalszynkiewicz) [reply extensions/resteasy-reactive/rest-client-reactive/runtime/src/main/java/io/quarkus/rest/client/reactive/runtime/MicroProfileRestClientRequestFilter.java:75]: that's the way the factory in the MP spec works. It takes in headers that come from an incoming request and creates a new map of headers from it. It's basically about propagating selected headers.
I'll add a comment about it. But I started to have some doubts re how this code works with calls that do blocking stuff so will first do some experiments :)

**Comment by michalszynkiewicz** [comment]
> Is this how is working the resteasy classic vendor of the reactive client header annotation?
> 
> If not, I'm not sure if this should be the right behavior. Mostly because of the scenario of using this annotation in sub-resources where the headers of the latest resource will take precedent.

the `incomingHeaders` are not the headers from previous annotations, etc but headers from a call that initiated this client call.
So, if e.g. curl calls our endpoint with a set of headers, then the endpoint uses the client to call another endpoint. `incomingHeadeders` will contain headers sent by curl.

**Comment by michalszynkiewicz** [comment]
By the way I also added the header params that are defined by the client as a parameter to this method. This is how it works in the classic variant. 

I had to change how the suspend/resume is done - it wasn't working for async invocations. I added the timer in the test to check for it. This test was failing before the change.

**Review by Sgitario — APPROVED** [review]

File: extensions/resteasy-reactive/rest-client-reactive/runtime/src/main/java/io/quarkus/rest/client/reactive/ReactiveClientHeadersFactory.java [hunk 2]
```diff
@@ -10,7 +10,20 @@
...
+     * on the outbound processing chain could further update the headers.
+     *
+     * @param incomingHeaders the map of headers from the inbound JAX-RS request. This will be an empty map if the
+     *        associated client interface is not part of a JAX-RS request.
+     * @param clientOutgoingHeaders the read-only map of header parameters specified on the client interface.
+     * @return a Uni with a map of HTTP headers to merge with the clientOutgoingHeaders to be sent to the remote service.
+     *
+     * @see ClientHeadersFactory#update(MultivaluedMap, MultivaluedMap)
+     */
+    public abstract Uni<MultivaluedMap<String, String>> getHeaders(MultivaluedMap<String, String> incomingHeaders,
```
Comment (geoand) [inline extensions/resteasy-reactive/rest-client-reactive/runtime/src/main/java/io/quarkus/rest/client/reactive/ReactiveClientHeadersFactory.java:25]: This is a breaking change, so we should add the appropriate label and ready an entry for the migration guide.

**Review by geoand — APPROVED** [review]

**Comment by michalszynkiewicz** [comment]
https://github.com/quarkusio/quarkus/wiki/Migration-Guide-2.8#rest-client-reactive-reactiveclientheadersfactory-method-change - migration guide

**Comment by michalszynkiewicz** [comment]
Without it the reactive factory is not much use for real blocking stuff so I'm wondering if even though the change is blocking, isn't it worth backporting it.

WDYT @geoand  @gsmet ?

**Comment by geoand** [comment]
I'm +0

**Pull request merged**